File: src/BacktrackingKP01wCG.cpp
// BacktrackingKP01.cpp
#include "BacktrackingKP01wCG.h"

BacktrackingKP01wCG::BacktrackingKP01wCG() {}

// Función auxiliar que calcula el beneficio máxima posible agregando todos los ítems de k en adelante
int BacktrackingKP01wCG::maxProfit(Solution S, int k){
    int n = S.numItems();

    for (int i = k; i<n; i++){
        S.addItem(i); //Agregamos todos los items de k en adelante ya que como no hay beneficios negativos, el beneficio maximo a partir de k implica agregar todos los items
    }

    return S.totalProfit(); //Devolvemos el beneficio maximo
}

void BacktrackingKP01wCG::solveAux(const KP01withCGInstance& mochila, int k, Solution& B, Solution S){
    int n = mochila.getNumItems();

    //Caso base: si ya recorrimos todos los items
    if(k == n){
        //Si el peso total de la posible solucion es menor o igual que la capacidad de la mochila, su beneficio es mayor que el de la mejor solucion encontrada hasta el momento y sus items no tienen conflictos
        if(S.totalWeight()<=mochila.getCapacity() && S.totalProfit()> B.totalProfit() && S.isValidItems()){
            //Se guarda la mejor solucion encontrada hasta el momento
            B=S;
        }
    } 
    // Caso recursivo: si la posible solucion es válida, no tiene conflictos y el beneficio potencial máximo de su sub-arbol es mayor que la mejor hasta ahora
    else if(S.isValidItems() && S.totalWeight()<=mochila.getCapacity() && maxProfit(S, k)>B.totalProfit()){
        Solution S2 = S; //Creamos una copia del estado actual de la solucion
        S2.addItem(k); //Agregamos el item k
        solveAux(mochila, k+1, B, S2); //Llamado recursivo incluyendo al item k
        S2.removeItem(k); //Sacamos k
        solveAux(mochila, k+1, B, S2); //Llamado recursivo sin incluir al item k
    }
}

Solution BacktrackingKP01wCG::solve(const KP01withCGInstance& instance) {
    Solution ret(1);  //Creamos una solucion vacia
    ret.setMochila(instance); // Asociamos la instancia del problema a la solución
    solveAux(instance, 0, ret, ret); // Iniciamos la recursión desde el ítem 0

    return ret;
}

File: src/BacktrackingKP01wCG.py
from KP01withCGInstance import KP01withCGInstance
from Solution import Solution
import os

#4
class BacktrackingKP01wCG:
    def __init__(self):
        # Constructor vacío
        pass

    def max_profit(self, S: Solution, k: int, mochila: KP01withCGInstance) -> int:
        # Calcula el beneficio maximo que se puede obtener a partir de una solucion parcial y los items que faltan 
        n = S.num_items()
        S2 = Solution(n)
        S2.set_mochila(mochila)
        S2._items_vector = S.get_items_vector().copy()

        # El beneficio maximo (sin tener en cuenta la factibilidad) implica agregar todos los items a partir de k por lo que los agregamos y vemos el beneficio final
        for i in range(k, n):
            S2.add_item(i)
        return S2.total_profit()

    def solve_aux(self, mochila: KP01withCGInstance, k: int, B: Solution, S: Solution):
        n = mochila.get_num_items()
        
        if k == n:
            # Caso base: ya vimos todos los ítems
            if (S.total_weight() <= mochila.get_capacity() and
                S.total_profit() > B.total_profit() and
                S.is_valid_items()):
                # Si S es mejor solución y válida, la guardamos como la mejor hasta ahora                
                B.set_mochila(mochila)
                B._items_vector = S.get_items_vector().copy()
        
        # Vemos si todavía vale la pena seguir teniendo en cuenta el beneficio potencial de ese sub arbol (poda por factibilidad y optimalidad)
        elif (S.is_valid_items() and S.total_weight() <= mochila.get_capacity() and
            self.max_profit(S, k, mochila) > B.total_profit()):
            S2 = Solution(n)
            S2.set_mochila(mochila)
            S2._items_vector = S.get_items_vector().copy()
            
            # Agregamos al item k y evaluamos la solucion con este
            S2.add_item(k) 
            self.solve_aux(mochila, k + 1, B, S2)
            
            #Sacamos al item k y evaluamos la solucion sin este
            S2.remove_item(k)
            self.solve_aux(mochila, k + 1, B, S2)

    def solve(self, instance: KP01withCGInstance) -> Solution:
        ret = Solution(1) # Inicializa solución vacía
        ret.set_mochila(instance)
        self.solve_aux(instance, 0, ret, ret) # Comienza el backtracking desde el ítem 0
        return ret
    
def test_backtracking():
    # Crear una instancia de la mochila desde un archivo
    instance = KP01withCGInstance(0, 0)
    instance.set_instance("instances/test_instance.in")

    # Crear el solucionador con backtracking
    bt_solver = BacktrackingKP01wCG()
    
    # Resolver la instancia
    solution = bt_solver.solve(instance)
    
    # Imprimir la solución
    print("Total Profit:", solution.total_profit())
    print("Total Weight:", solution.total_weight())
    print("Selected Items:", solution.get_items_vector())
    print("Capacity:", instance.get_capacity())
    
    # Validar que la solución es válida
    assert solution.total_weight() <= instance.get_capacity(), "Error: Se superó la capacidad de la mochila"
    assert solution.is_valid_items(), "Error: La solución tiene elementos en conflicto"
    print("Test passed successfully!")

if __name__ == "__main__":
    file_path = input("Enter the path to the instance file: ").strip()
    if not os.path.exists(file_path):
        print(f"Error: File '{file_path}' does not exist.")
        exit(1)
    instance = KP01withCGInstance(0, 0)
    instance.set_instance(file_path)

    bt_solver = BacktrackingKP01wCG()
    solution = bt_solver.solve(instance)
    print("Weight used: ", solution.total_weight())
    print("Profit obtained: ", solution.total_profit())

    solution.print_solution()